### PR TITLE
[Data Explorer][Discover 2.0] Fix issues when change index pattern

### PR DESCRIPTION
**Pull request by ananzh**
### Description
* side nav only shows the selected index pattern fields after switch index pattern
Screenshot 2023-08-31 at 09 23 24

* allow reset column state when switch index pattern

There are two conditions: 1) If there is common field, common field will be saved. 
Screenshot 2023-08-31 at 09 23 51

Screenshot 2023-08-31 at 09 23 59

2) if no common field, will show `_source` after switch
Screenshot 2023-08-31 at 09 23 34

Screenshot 2023-08-31 at 09 23 43

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4840 https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4846

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/discover/public/application/components/sidebar/lib/get_index_pattern_field_list.ts [hunk 1]
```diff
@@ -36,18 +36,5 @@ export function getIndexPatternFieldList(
...
-
-  difference(fieldNamesInDocs, fieldNamesInIndexPattern).forEach((unknownFieldName) => {
-    unknownTypes.push({
-      displayName: String(unknownFieldName),
-      name: String(unknownFieldName),
-      type: 'unknown',
-    } as IndexPatternField);
-  });
-
-  return [...indexPattern.fields.getAll(), ...unknownTypes];
```
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/sidebar/lib/get_index_pattern_field_list.ts:52]: Why did you remove this?

File: src/plugins/discover/public/application/view_components/utils/filter_columns.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { IndexPattern } from '../../../opensearch_dashboards_services';
+
+/**
+ * Helper function to filter columns based on the fields of the index pattern.
+ * This function is used when we switch between index patterns. We want to keep the columns that are
+ * still available in the new index pattern and remove the ones that are not.
+ * If the resulting array is empty, it provides a fallback to a single '_source' column.
```
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/utils/filter_columns.ts:12]: We should use the default column from advanced settings and not hardcode `_source`. `DEFAULT_COLUMNS_SETTING`

File: src/plugins/discover/public/application/view_components/canvas/index.tsx [hunk 3]
```diff
@@ -14,10 +14,18 @@ import { ResultStatus, SearchData } from '../utils/use_search';
...
+import { filterColumns } from '../utils/filter_columns';
 
 // eslint-disable-next-line import/no-default-export
 export default function DiscoverCanvas({ setHeaderActionMenu, history }: ViewProps) {
   const { data$, refetch$, indexPattern } = useDiscoverContext();
+  const { columns } = useSelector((state) => state.discover);
+  const filteredColumns = filterColumns(columns, indexPattern);
+  const dispatch = useDispatch();
+
+  const prevColumns = useRef(filteredColumns);
```
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/index.tsx:27]: Doesn't look like we are using this. Do we need it?

File: src/plugins/discover/public/application/view_components/utils/filter_columns.ts (hunk 2 above)
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/utils/filter_columns.ts:12]: will update

File: src/plugins/discover/public/application/view_components/canvas/index.tsx (hunk 3 above)
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/canvas/index.tsx]: will remove.

File: src/plugins/discover/public/application/components/sidebar/lib/get_index_pattern_field_list.ts (hunk 1 above)
Comment (ananzh) [reply src/plugins/discover/public/application/components/sidebar/lib/get_index_pattern_field_list.ts:52]: fieldNamesInDocs has all the fields from prev index pattern. This difference will compare fields and add prev idx fields to unknownTypes and return them. This is why we see unknownTypes (grey fields) when we switch index pattern. I think we agree that for discover, we only need selected index pattern fields for now.

Comment (ashwin-pc) [reply src/plugins/discover/public/application/components/sidebar/lib/get_index_pattern_field_list.ts:52]: Fair point :)

**Review by ashwin-pc — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]
